dnsprovider/pkg/dnsprovider/providers/scaleway/dns_test.go

```diff
@@ -0,0 +1,301 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dns
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/google/uuid"
+	domain "github.com/scaleway/scaleway-sdk-go/api/domain/v2beta1"
+	"k8s.io/kops/cloudmock/scaleway/mockdns"
+	"k8s.io/kops/dnsprovider/pkg/dnsprovider"
+	"k8s.io/kops/dnsprovider/pkg/dnsprovider/rrstype"
+)
+
+func setUpFakeZones() *mockdns.FakeDomainAPI {
+	return &mockdns.FakeDomainAPI{
+		DNSZones: []*domain.DNSZone{
+			{
+				Domain:    "test.fr",
+				Subdomain: "zone",
+			},
+			{
+				Domain:    "test.fr",
+				Subdomain: "kops",
+			},
+			{
+				Domain:    "domain.fr",
+				Subdomain: "zone",
+			},
+		},
+	}
+}
+
+func setUpFakeZonesNil() *mockdns.FakeDomainAPI {
+	return &mockdns.FakeDomainAPI{}
+}
+
+func getDNSProviderZones(api DomainAPI) dnsprovider.Zones {
+	dnsProvider := NewProvider(api)
+	zs, _ := dnsProvider.Zones()
+	return zs
+}
+
+func TestZonesListValid(t *testing.T) {
+	domainAPI := setUpFakeZones()
+	z := &zones{domainAPI: domainAPI}
+
+	zoneList, err := z.List()
+
+	if err != nil {
+		t.Errorf("error listing zones: %v", err)
+	}
+	if len(zoneList) != 3 {
+		t.Errorf("expected at least 1 zone, got 0")
+	}
+	for i, zone := range zoneList {
+		if zone.Name() != domainAPI.DNSZones[i].Domain {
+			t.Errorf("expected %s as zone name, got: %s", domainAPI.DNSZones[i].Domain, zone.Name())
+		}
+	}
+}
+
+func TestZonesListShouldFail(t *testing.T) {
+	domainAPI := setUpFakeZonesNil()
+	z := &zones{domainAPI: domainAPI}
+
+	zoneList, err := z.List()
+
+	if err == nil {
+		t.Errorf("expected non-nil err")
+	}
+	if zoneList != nil {
+		t.Errorf("expected nil zone, got %v", zoneList)
+	}
+}
+
+func TestAddValid(t *testing.T) {
+	domainAPI := setUpFakeZones()
+	zs := getDNSProviderZones(domainAPI)
+
+	inZone := &zone{
+		name:      "dns.test.fr",
+		domainAPI: domainAPI,
+	}
+	outZone, err := zs.Add(inZone)
+
+	if err != nil {
+		t.Errorf("unexpected err: %v", err)
+	}
+	if outZone == nil {
+		t.Errorf("zone is nil, exiting test early")
+	}
+	if outZone.Name() != "dns" {
+		t.Errorf("unexpected zone name: %s", outZone.Name())
+	}
+}
+
+func TestAddShouldFail(t *testing.T) {
+	domainAPI := setUpFakeZonesNil()
+	zs := getDNSProviderZones(domainAPI)
+
+	inZone := &zone{
+		name:      "dns.test.fr",
```

Comment:
Nit: We normally use example.com - it's reserved for tests/examples: https://en.wikipedia.org/wiki/Example.com

(Maybe test.fr is also though?)

Not a blocker!